New sidenav

jonasbrinkhoff opened the pull request:
**New: Updated the behaviour of the side nav based on Toms suggestion**

Sub-sections and sub-pages now open freely while on a page, but then only the page you're currently in is open when you actually visit a page.

**Summary of changes:**
- Updated the design of the sidenav
- Separated the content and logic of the navigation. Navigation tree can now be changed in `/data/navigation.yml`
- Renamed a couple of files so that their active state is properly set by the `path_active` function in `config.rb` file

tombruijn commented on source/_side_nav.html.erb:
```diff
@@ -0,0 +1,906 @@
+<%
+  sections = [
```
if we're going to define it like this, we should probably move it to a YAML file. Should be easier to edit.

jonasbrinkhoff replied:
Done :)

tombruijn commented on config.rb:
```diff
@@ -77,6 +77,14 @@ def site_name
+      return current_page.url === '/'
+    end
+
+    return current_page.url.include?(path)
```
```suggestion
    current_page.url.include?(path)
```
Unnecessary return

tombruijn commented on source/_side_nav.html.erb:
```diff
@@ -0,0 +1,906 @@
+
+<nav>
+  <ul>
+    <% sections.each do |section| %>
```
This whole thing seems really complex.

I don't like how I need to first click a section to wait for a page load and then only see the section links. That means it takes me X amount of clicks with page loads to go to the page I need to visit. Makes the navigation slower.

I'd rather keep the open and close logic in JavaScript.

jonasbrinkhoff replied:
I talked with Wes about this (and I even implemented the old navigation style), but we came to the conclusion to keep it like it is now.

With the old navigation style, you either end up with a really long navigation (because multiple sub-navigations are open) or you'd loose the context of where you currently are (because once you opened another sub-navigation, the currently active one would close).

We are aware that the new solution increases the amount of clicks and time to navigate through the page, but we think that the drawbacks of the other solutions are bigger. In the long term, we'd like to move the documentation to Gatsby (or something similar) which would drastically reduce page load times and thereby hopefully eliminate the problems with the new navigation.

tombruijn replied:
> With the old navigation style, you either end up with a really long navigation (because multiple sub-navigations are open) or you'd loose the context of where you currently are (because once you opened another sub-navigation, the currently active one would close).

There are other solutions to that problem as discussed. Another idea is to freely open sections while on a page, but then only the section you're in is open when you actually visit a page.

> In the long term, we'd like to move the documentation to Gatsby (or something similar) which would drastically reduce page load times and thereby hopefully eliminate the problems with the new navigation.

Load times are not the issue. As a user, on the deployed site, you're still waiting for page load times. That's my problem.

jonasbrinkhoff replied:
> Another idea is to freely open sections while on a page, but then only the section you're in is open when you actually visit a page.

Really like this idea, implement it :)

jonasbrinkhoff commented:
@wesoudshoorn 

> The links are now rendered directly inside an, which is unvalid HTML

I think something is missing here in this comment?

> We used to have a single side navigation and hid subitems with JS, we now have everything hidden based on Ruby

Updated the code, non-active sub-navigation elements are now rendered but hidden.

> Metrics and standalone agent links do not work

Fixed all links now

jonasbrinkhoff commented:
Updated all page names & links now, all links should now work properly in the navigation.

Also refactored the navigation a bit to make it more readable.

wesoudshoorn commented:
@jonasbrinkhoff https://deploy-preview-344--dreamy-wright-e9a95c.netlify.app/ the deploy preview is broken now?

wesoudshoorn commented on data/navigation.yml:
```diff
@@ -0,0 +1,631 @@
+    name: "Features"
+    sub_sections:
+      -
+        name: "Standalone agent<sup>Beta</sup>"
```
```suggestion
        name: "Standalone agent <sup>Beta</sup>"
```

jonasbrinkhoff commented:
Updated the behaviour of the side nav based on Toms suggestion and moved around some pages to make it work. Would be great if you could review it again :)

jvanbaarsen commented:
@jonasbrinkhoff What is the status of this? Only waiting for reviews?

jonasbrinkhoff commented:
> @jonasbrinkhoff What is the status of this? Only waiting for reviews?

Yes, only waiting for reviews!

tombruijn commented on data/navigation.yml:
```diff
@@ -0,0 +1,697 @@
+                link: "/application/settings/application-settings.html"
+              -
+                name: "Notification settings"
+                link: "/application/settings/notification-settings.html"
```
I wouldn't want to move pages in this PR. Yes it's now matching the new "navigation structure", but we're going to restructure it later in the docs improvement project. We then need to move it again.

The problem being that moving files require redirects to not break existing links. (This PR doesn't update those redirects as defined in `netlify.toml`.) So in this PR we would need to add more redirects, then later in the docs improvements projects we need to add even more redirects. And by then we've exhausted netlify's redirection list, if they still only support a limited amount of redirects.

jonasbrinkhoff commented on config.rb:
```diff
@@ -77,6 +77,14 @@ def site_name
   end
 
+  def path_active?(path)
+    if path === '/'
+      return current_page.url === '/'
```
`current_page.url.include?("/")` would be true for all paths. That's why, if the given `path` is `"/"`, it should only return true if the `current_path` is also `"/"`.

tombruijn replied:
Yes, but that would also be covered under `path == current_path.url`. That should then also work for most paths that we can match 1 to 1 and don't have to check for a partial match with `include?`/`start_with?`.

jonasbrinkhoff replied:
I'm not sure if I got your comment. But we need a partial check, to determine the active state of parent elements.

Here's how `path_active` should work like, maybe this makes things more clear:
- when on `"/"`
  - `path_active("/")` should be `true`
  - `path_active("/application/")` should be `false`
- when on `"/application/"`
  - `path_active("/")` should be `false` <- weird case
  - `path_active("/application/")` should be `true`
  - `path_active("/application/index.html")` should be `true` <- weird case (only exists if we use index.html)
  - `path_active("/application/new-application.html")` should be `false`

jonasbrinkhoff commented on config.rb:
```diff
@@ -77,6 +77,14 @@ def site_name
+      return current_page.url === '/'
+    end
+
+    current_page.url.include?(path)
```
Oh, good point! Didn't knew `start_with?` exists. I'll update it :)

jonasbrinkhoff commented on data/navigation.yml:
```diff
@@ -0,0 +1,697 @@
+        pages:
+          -
+            name: "Overview"
+            link: "/application/overview.html"
```
I generally agree, but this would break the `path_active` function because of how `current_page.url` works.

Example:
- You rename `/application/overview.html` file to `/application/index.html`
- You then visit https://docs.appsignal.com/application/index.html
- You'd expect `current_page.url` to return `/application/index.html`, but in reality it returns just `/application/`
- And since `"/application/".start_with?("/application/index.html")` is `false`, `path_active` would return `false`

If you have any idea how to fix this, let me know.

tombruijn commented:
@jonasbrinkhoff I've renamed all the overview.html to index.html and updated the logic in the last commit. Can you let me know if it works good now?

tombruijn commented:
I'll have a look what pages should be renamed back to their originals. I don't think the pages under "Your application > Settings" should really go there. I've put them there to give some form of structure, but every page upon themself is a feature  we should put in a better structure in the future.

tombruijn commented:
@jonasbrinkhoff I've redirected and reverted some files, see last commit. I think it's fine to merge now.

jonasbrinkhoff commented:
> @jonasbrinkhoff I've redirected and reverted some files, see last commit. I think it's fine to merge now.

Amazing, thanks for your help!
